web3/middleware/cache.py

```diff
@@ -0,0 +1,328 @@
+import time
+
+import lru
+
+from web3.utils.caching import (
+    generate_cache_key,
+)
+
+
+SIMPLE_CACHE_RPC_WHITELIST = {
+    'web3_clientVersion',
```

Comment:
What happens if someone connects to a new provider? I don't think we can do a simple cache on `web3_clientVersion`.

Reply:
I'm going to make some changes that will cause the cache to get busted anytime the middleware stack is rebuilt (which I believe happens anytime providers change).